// theme-ui/dist/theme-ui.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@theme-ui/core');
var colorModes = require('@theme-ui/color-modes');
var themeProvider = require('@theme-ui/theme-provider');
var components = require('@theme-ui/components');
var css = require('@theme-ui/css');

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var BaseStyles = function BaseStyles(props) {
  return jsx('div', _extends({}, props, {
    sx: _extends({
      fontFamily: 'body',
      lineHeight: 'body',
      fontWeight: 'body',
      variant: 'styles'
    }, props.sx)
  }));
};
var jsx = core.jsx;

Object.defineProperty(exports, '__ThemeUIContext', {
  enumerable: true,
  get: function () { return core.__ThemeUIContext; }
});
Object.defineProperty(exports, 'createElement', {
  enumerable: true,
  get: function () { return core.createElement; }
});
Object.defineProperty(exports, 'merge', {
  enumerable: true,
  get: function () { return core.merge; }
});
Object.defineProperty(exports, 'useThemeUI', {
  enumerable: true,
  get: function () { return core.useThemeUI; }
});
Object.defineProperty(exports, 'InitializeColorMode', {
  enumerable: true,
  get: function () { return colorModes.InitializeColorMode; }
});
Object.defineProperty(exports, 'useColorMode', {
  enumerable: true,
  get: function () { return colorModes.useColorMode; }
});
Object.defineProperty(exports, 'ThemeProvider', {
  enumerable: true,
  get: function () { return themeProvider.ThemeProvider; }
});
Object.defineProperty(exports, 'css', {
  enumerable: true,
  get: function () { return css.css; }
});
Object.defineProperty(exports, 'get', {
  enumerable: true,
  get: function () { return css.get; }
});
exports.BaseStyles = BaseStyles;
exports.jsx = jsx;
Object.keys(components).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return components[k]; }
  });
});
